lib/internal/webstreams/readablestream.js

```diff
@@ -477,9 +477,12 @@ class ReadableStream {
 
     // eslint-disable-next-line no-use-before-define
     const reader = new ReadableStreamDefaultReader(this);
-    let done = false;
+
+    const state = {
```

Comment:
Further improvement potential - we don't need the extra object here necessarily

Reply:
definetly a perf bottle neck because `{__proto__: null}` super slow is.

Reply:
Benchmark ci with proto: https://ci.nodejs.org/job/benchmark-node-micro-benchmarks/1394/

Reply:
With proto:

```
confidence improvement accuracy (*)   (**)  (***)
webstreams/pipe-to.js highWaterMarkW=1024 highWaterMarkR=1024 n=500000        ***     62.47 %       ±3.03% ±4.06% ±5.33%
webstreams/pipe-to.js highWaterMarkW=1024 highWaterMarkR=2048 n=500000        ***     66.26 %       ±4.13% ±5.55% ±7.32%
webstreams/pipe-to.js highWaterMarkW=1024 highWaterMarkR=4096 n=500000        ***     63.66 %       ±3.79% ±5.08% ±6.68%
webstreams/pipe-to.js highWaterMarkW=1024 highWaterMarkR=512 n=500000         ***     60.67 %       ±3.29% ±4.40% ±5.78%
webstreams/pipe-to.js highWaterMarkW=2048 highWaterMarkR=1024 n=500000        ***     62.22 %       ±3.94% ±5.28% ±6.94%
webstreams/pipe-to.js highWaterMarkW=2048 highWaterMarkR=2048 n=500000        ***     65.02 %       ±3.82% ±5.13% ±6.77%
webstreams/pipe-to.js highWaterMarkW=2048 highWaterMarkR=4096 n=500000        ***     64.40 %       ±4.17% ±5.61% ±7.42%
webstreams/pipe-to.js highWaterMarkW=2048 highWaterMarkR=512 n=500000         ***     63.58 %       ±3.90% ±5.23% ±6.89%
webstreams/pipe-to.js highWaterMarkW=4096 highWaterMarkR=1024 n=500000        ***     63.92 %       ±4.16% ±5.57% ±7.34%
webstreams/pipe-to.js highWaterMarkW=4096 highWaterMarkR=2048 n=500000        ***     61.79 %       ±3.87% ±5.19% ±6.84%
webstreams/pipe-to.js highWaterMarkW=4096 highWaterMarkR=4096 n=500000        ***     64.32 %       ±3.51% ±4.70% ±6.18%
webstreams/pipe-to.js highWaterMarkW=4096 highWaterMarkR=512 n=500000         ***     64.31 %       ±4.11% ±5.52% ±7.29%
webstreams/pipe-to.js highWaterMarkW=512 highWaterMarkR=1024 n=500000         ***     62.24 %       ±3.70% ±4.95% ±6.51%
webstreams/pipe-to.js highWaterMarkW=512 highWaterMarkR=2048 n=500000         ***     64.11 %       ±3.96% ±5.29% ±6.94%
webstreams/pipe-to.js highWaterMarkW=512 highWaterMarkR=4096 n=500000         ***     65.23 %       ±3.60% ±4.83% ±6.37%
webstreams/pipe-to.js highWaterMarkW=512 highWaterMarkR=512 n=500000          ***     59.86 %       ±3.51% ±4.69% ±6.15%
Be aware that when doing many comparisons the risk of a false-positive
result increases. In this case, there are 16 comparisons, you can thus
expect the following amount of false-positive results:
0.80 false positives, when considering a   5% risk acceptance (*, **, ***),
0.16 false positives, when considering a   1% risk acceptance (**, ***),
0.02 false positives, when considering a 0.1% risk acceptance (***)
```

Reply:
with `__proto__: null` the results are (from 140% to 120%):
```
                                                                       confidence improvement accuracy (*)    (**)   (***)
webstreams/readable-async-iterator.js n=100000                                ***    120.32 %      ±13.64% ±18.28% ±24.06%

Be aware that when doing many comparisons the risk of a false-positive
result increases. In this case, there are 23 comparisons, you can thus
expect the following amount of false-positive results:
  1.15 false positives, when considering a   5% risk acceptance (*, **, ***),
  0.23 false positives, when considering a   1% risk acceptance (**, ***),
  0.02 false positives, when considering a 0.1% risk acceptance (***)
```

Reply:
this was stated before - we don't really need a state object.

Reply:
@MoLow Raz explained to me in DM we do because we mutate it from within a class and this avoids a closure

Reply:
Do we really need to have NullObject?

If so, we could do it like this

```
const ReadState = function ReadState () {
  this.done = false;
  this.current = undefined;
}
ReadState.prototype = Object.create(null)
```

not tested ..

Reply:
Or `class ReadState extends null`?

Reply:
(not identical but same purpose)

Reply:
Is this notation possible? Never tested and benched it.

Reply:
`extends null` does not work

when missing super:
```
ReferenceError: Must call super constructor in derived class before accessing 'this' or returning from derived constructor
```

when adding super:
```
TypeError: Super constructor null of ReadState is not a constructor
```

Reply:
Then the prototypical way.

Btw. What do you mean it adds a scope? Could you please explain how the local variable creates an additional scope?

Reply:
extracting outside the prototype way would further complicate the code and I don't think it's worth it...

Reply:
Could you still point me to a document regarding the additional scope thing and why it is bad?

Reply:
which scope?

do you mean closure?

Reply:
Ah, yes closure. I dont get why an object is ok but primitive types not.

Reply:
I'm not 100% sure from what the improvement came from (it was playing around and trying different things) but one of the big possibilities is that there is no need to hold on to the parent closure variables when it's in a different scope completely

Reply:
@Uzlopak this is quite outdated but worth a read https://mrale.ph/blog/2012/09/23/grokking-v8-closures-for-fun.html